internal/nvmeof/controller/controllerserver.go

```diff
@@ -0,0 +1,526 @@
+		"nvmeof.GatewayPort":    nvmeofData.GatewayManagementInfo.Port,
+	}
+
+	// Store all metadata entries
+	for key, value := range metadata {
+		if value == "" {
+			log.WarningLog(ctx, "Skipping empty metadata value for key: %s", key)
+
+			continue
+		}
+
+		if err := rbdVol.SetMetadata(key, value); err != nil {
+			log.ErrorLog(ctx, "Failed to store %s metadata: %v", key, err)
+
+			return status.Errorf(codes.Internal, "failed to store %s metadata: %v", key, err)
+		}
+	}
+
+	log.DebugLog(ctx, "All NVMe-oF metadata stored successfully for volume: %s", volumeID)
+
+	return nil
+}
+
+// getNVMeoFMetadata retrieves all NVMe-oF data from RBD volume metadata.
+func (cs *Server) getNVMeoFMetadata(
+	ctx context.Context,
+	req *csi.DeleteVolumeRequest,
+	volumeID string,
+) (*nvmeof.NVMeoFVolumeData, error) {
+	// Create RBD manager
+	mgr := rbdutil.NewManager(cs.backendServer.Driver.GetInstanceID(), nil, req.GetSecrets())
+	defer mgr.Destroy(ctx)
+
+	// Get RBD volume
+	rbdVol, err := mgr.GetVolumeByID(ctx, volumeID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find volume with ID %q: %w", volumeID, err)
+	}
+	defer rbdVol.Destroy(ctx)
+
+	// Retrieve all metadata including gateway management info
+	metadataKeys := map[string]*string{
+		"nvmeof.SubsystemNQN":     new(string),
+		"nvmeof.NamespaceID":      new(string),
+		"nvmeof.HostNQN":          new(string),
+		"nvmeof.ListenerAddress":  new(string),
+		"nvmeof.ListenerPort":     new(string),
+		"nvmeof.ListenerHostname": new(string),
+		"nvmeof.GatewayAddress":   new(string),
+		"nvmeof.GatewayPort":      new(string),
+	}
+
+	// Retrieve all metadata values
+	for key, valuePtr := range metadataKeys {
+		value, err := rbdVol.GetMetadata(key)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get %s: %w", key, err)
+		}
+		*valuePtr = value
+	}
+
+	// Parse namespace ID
+	nsid, err := strconv.ParseUint(*metadataKeys["nvmeof.NamespaceID"], 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("invalid namespace ID: %w", err)
+	}
+
+	// Construct NVMe-oF volume data
+	nvmeofData := &nvmeof.NVMeoFVolumeData{
+		SubsystemNQN: *metadataKeys["nvmeof.SubsystemNQN"],
+		NamespaceID:  uint32(nsid),
+		HostNQN:      *metadataKeys["nvmeof.HostNQN"],
+		ListenerInfo: nvmeof.ListenerDetails{
+			GatewayAddress: nvmeof.GatewayAddress{
+				Address: *metadataKeys["nvmeof.ListenerAddress"],
+				Port:    *metadataKeys["nvmeof.ListenerPort"],
+			},
+			Hostname: *metadataKeys["nvmeof.ListenerHostname"],
+		},
+		// Store gateway management info separately
+		GatewayManagementInfo: nvmeof.GatewayAddress{
+			Address: *metadataKeys["nvmeof.GatewayAddress"],
+			Port:    *metadataKeys["nvmeof.GatewayPort"],
+		},
+	}
+
+	return nvmeofData, nil
+}
+
+// getGatewayConfigFromRequest extracts gateway configuration from request parameters.
+func (cs *Server) getGatewayConfigFromRequest(req *csi.CreateVolumeRequest) (*nvmeof.GatewayConfig, error) {
```

Comment:
Functions that do not use `cs *Server` should not be a function of the struct either, just make it a simple stand-alone helper function.